Cabal/src/Distribution/Simple/BuildPaths.hs

```diff
@@ -82,10 +88,30 @@ buildInfoPref distPref = distPref </> "build-info.json"
 
 -- | This is the name of the directory in which the generated haddocks
 -- should be stored. It does not include the @<dist>/doc/html@ prefix.
+--
+-- It is also used by `haddock-project` when constructing its output directory.
 haddockDirName :: HaddockTarget -> PackageDescription -> FilePath
 haddockDirName ForDevelopment = prettyShow . packageName
 haddockDirName ForHackage = (++ "-docs") . prettyShow . packageId
 
+-- | This is the name of the directory in which the generated haddocks for
+-- a (sub)library should be stored. It does not include the @<dist>/doc/html@
+-- prefix.
+--
+-- It is also used by `haddock-project` when constructing its output directory.
+haddockLibraryDirName
+  :: HaddockTarget
+  -> PackageDescription
+  -> Library
+  -> FilePath
+haddockLibraryDirName ForDevelopment pkg_descr lib =
+  case libName lib of
+    LSubLibName sublib_name ->
+      prettyShow (packageName pkg_descr) ++ "-" ++ prettyShow sublib_name
+    _ -> prettyShow (packageName pkg_descr)
+-- TODO: how Hackage should deal with sublibraries?
```

Comment:
How do we resolve this question? Do you want to ask gbaz?

Reply:
Yes, this needs to be considered.

I see two choices:

* each sublibrary is available from it's own page;
* All public sublibraries are available on the main library page

The former might be cumbersome (since making a revision of each sublibrary will need to make a revision of all of them).

For the latter one might be able to use `cabal haddock-project --hackage` maybe with some modifications to write a directory containing haddocks of all sublibraries which can be served by hackage.  This might be not only more ergonomic but also the simplest way to support public sublibraries on hackage.

_desclaimer: I have never contributed to `hackage` so please take my word with a grain of salt :smile:_.

Reply:
^ ping @gbaz

Reply:
> The former might be cumbersome (since making a revision of each sublibrary will need to make a revision of all of them).

What do you mean by "revision" in this sentence? A cabal file revision on Hackage is a package-level thing so it would affect all package's libraries.

Reply:
I mean package revisions Hackage.  I was considering if a sublibrary is exposed on its own, e.g. `hackage.com/package/mypackage-sublibrary` while the main library is served from `hackage.com/package/mypackage`.

But if we publish sublibraries as part of the package, e.g. `hackage.com/package/mypackage/sublibrary` then it will be more consistent.  Also finding sublibraries of a package will be simpler if links to them will be rendered on the package page.

Reply:
I think `mypackage/sublibrary` is the correct option here.

Reply:
@coot cool, I'd love this to be documented somewhere so that we don't forget it exists

Reply:
I adjusted the paths, but there's more work either on `cabal` side or `hackage` side. If we have a `LIB` with a sublibrary `SUBLIB`, then `LIB` docs will be placed under `LIB-0.1.0.0/doc/html/LIB` and the documentation of the sublibrary will be placed under `LIB-0.1.0.0/l/SUBLIB/doc/html/LIB/SUBLIB/`.

There are various options from here to get sublibrary support on hackage:
- use `haddock-project` on hackage - this might require some changes to `haddock-project`;
- copy the sublibrary documentation to the right place;
- as above, but let `cabal` do that.  This might require changing the directory layout of `dist-newstyle` to be consistent (e.g. drop the `l/t` subdirs).

I think the first option is the best one - but this is outside of this PR anyway.